File: programs/clear-wallet/src/utils/definition.rs
use quasar_lang::prelude::*;

// --- Enums ---
// All #[repr(u8)] so they're 1 byte, alignment 1 — safe in #[repr(C)] structs.

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ParamType {
    Address = 0,
    U64 = 1,
    I64 = 2,
    String = 3,
    Bool = 4,
    U8 = 5,
    U16 = 6,
    U32 = 7,
    U128 = 8,
    /// 20-byte fixed buffer. Used for EVM addresses and Bitcoin HASH160
    /// (P2WPKH witness program).
    Bytes20 = 9,
    /// 32-byte fixed buffer. Used for transaction hashes, BTC scriptPubKey
    /// hashes (P2WSH/P2TR), and large counters.
    Bytes32 = 10,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConstraintType {
    None = 0,
    LessThanU64 = 1,
    GreaterThanU64 = 2,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AccountSourceType {
    Static = 0,
    Param = 1,
    PdaDerived = 2,
    HasOne = 3,
    Vault = 4,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SegmentType {
    Literal = 0,
    Param = 1,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SeedType {
    Literal = 0,
    ParamRef = 1,
    AccountRef = 2,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DataEncoding {
    RawAddress = 0,
    LittleEndianU64 = 1,
    LittleEndianI64 = 2,
    Bool = 3,
    LittleEndianU8 = 4,
    LittleEndianU16 = 5,
    LittleEndianU32 = 6,
    LittleEndianU128 = 7,
}

// --- Fixed-size entry types for Intent Vecs ---
//
// All #[repr(C)] with alignment-1 fields (u8, enums, PodU16/PodU64).
// PodU16/PodU64 are quasar's alignment-1 integer wrappers — they store
// multi-byte integers as [u8; N] so the struct stays alignment 1,
// which is required for quasar's zero-copy pointer casting.
//
// Variable-length or variant-specific data lives in the byte_pool.
// Each entry has a type discriminator and pool_offset + pool_len
// pointing to its data in the pool.

#[repr(C)]
#[derive(Clone, Copy)]
pub struct ParamEntry {
    pub param_type: ParamType,
    /// Byte pool offset + length of the param name (UTF-8).
    pub name_offset: PodU16,
    pub name_len: PodU16,
    pub constraint_type: ConstraintType,
    /// Constraint value (only meaningful when constraint_type != None).
    pub constraint_value: PodU64,
}

/// Account source entry. Pool data layout depends on source_type:
///   Static:     [address: 32 bytes]
///   Param:      [param_index: 1 byte]
///   PdaDerived: [program_account_index: 1, seeds_start: 2 LE, seeds_count: 2 LE]
///   HasOne:     [account_index: 1, byte_offset: 2 LE]
///   Vault:      (empty, pool_len = 0)
#[repr(C)]
#[derive(Clone, Copy)]
pub struct AccountEntry {
    pub is_signer: bool,
    pub is_writable: bool,
    pub source_type: AccountSourceType,
    pub pool_offset: PodU16,
    pub pool_len: PodU16,
}

#[repr(C)]
#[derive(Clone, Copy)]
pub struct InstructionEntry {
    pub program_account_index: u8,
    /// Byte pool range containing one account index byte per account.
    pub account_indexes_offset: PodU16,
    pub account_indexes_len: PodU16,
    /// Index range into the data_segments Vec.
    pub segments_start: PodU16,
    pub segments_count: PodU16,
}

/// Data segment entry. Pool data layout depends on segment_type:
///   Literal: [raw bytes]
///   Param:   [param_index: 1, encoding: 1]
#[repr(C)]
#[derive(Clone, Copy)]
pub struct DataSegmentEntry {
    pub segment_type: SegmentType,
    pub pool_offset: PodU16,
    pub pool_len: PodU16,
}

/// PDA seed entry. Pool data layout depends on seed_type:
///   Literal:    [raw bytes]
///   ParamRef:   [param_index: 1]
///   AccountRef: [account_index: 1]
#[repr(C)]
#[derive(Clone, Copy)]
pub struct SeedEntry {
    pub seed_type: SeedType,
    pub pool_offset: PodU16,
    pub pool_len: PodU16,
}

// --- Helpers ---

pub fn param_byte_size(param_type: ParamType, params_data: &[u8], offset: usize) -> Result<usize, ProgramError> {
    match param_type {
        ParamType::Address | ParamType::Bytes32 => Ok(32),
        ParamType::U64 | ParamType::I64 => Ok(8),
        ParamType::Bytes20 => Ok(20),
        ParamType::String => {
            let len = *params_data.get(offset).ok_or(ProgramError::InvalidInstructionData)? as usize;
            Ok(1 + len)
        }
        ParamType::Bool | ParamType::U8 => Ok(1),
        ParamType::U16 => Ok(2),
        ParamType::U32 => Ok(4),
        ParamType::U128 => Ok(16),
    }
}

impl DataEncoding {
    pub fn from_u8(v: u8) -> Option<Self> {
        match v {
            0 => Some(Self::RawAddress),
            1 => Some(Self::LittleEndianU64),
            2 => Some(Self::LittleEndianI64),
            3 => Some(Self::Bool),
            4 => Some(Self::LittleEndianU8),
            5 => Some(Self::LittleEndianU16),
            6 => Some(Self::LittleEndianU32),
            7 => Some(Self::LittleEndianU128),
            _ => None,
        }
    }

    pub fn byte_size(self) -> usize {
        match self {
            Self::RawAddress => 32,
            Self::LittleEndianU64 | Self::LittleEndianI64 => 8,
            Self::Bool | Self::LittleEndianU8 => 1,
            Self::LittleEndianU16 => 2,
            Self::LittleEndianU32 => 4,
            Self::LittleEndianU128 => 16,
        }
    }
}

pub fn encoding_byte_size(encoding: DataEncoding) -> usize {
    encoding.byte_size()
}

File: frontend/src/lib/msig/definition.ts
// Shared enums + small record types mirroring
// programs/clear-wallet/src/utils/definition.rs.
//
// All #[repr(u8)] on the Rust side . so the on-chain byte values must
// match these numeric constants exactly. We use const objects (not TS
// enums) because TS enums emit a runtime object with both string and
// numeric keys, which confuses byte-level parsers.

export const ParamType = {
  Address: 0,
  U64: 1,
  I64: 2,
  String: 3,
  Bool: 4,
  U8: 5,
  U16: 6,
  U32: 7,
  U128: 8,
  Bytes20: 9,
  Bytes32: 10,
  Bytes: 11,
} as const;
export type ParamType = (typeof ParamType)[keyof typeof ParamType];

export const ConstraintType = {
  None: 0,
  LessThanU64: 1,
  GreaterThanU64: 2,
} as const;
export type ConstraintType = (typeof ConstraintType)[keyof typeof ConstraintType];

export const AccountSourceType = {
  Static: 0,
  Param: 1,
  PdaDerived: 2,
  HasOne: 3,
  Vault: 4,
} as const;
export type AccountSourceType = (typeof AccountSourceType)[keyof typeof AccountSourceType];

export const SegmentType = {
  Literal: 0,
  Param: 1,
} as const;
export type SegmentType = (typeof SegmentType)[keyof typeof SegmentType];

export const SeedType = {
  Literal: 0,
  ParamRef: 1,
  AccountRef: 2,
} as const;
export type SeedType = (typeof SeedType)[keyof typeof SeedType];

export const DataEncoding = {
  RawAddress: 0,
  LittleEndianU64: 1,
  LittleEndianI64: 2,
  Bool: 3,
  LittleEndianU8: 4,
  LittleEndianU16: 5,
  LittleEndianU32: 6,
  LittleEndianU128: 7,
} as const;
export type DataEncoding = (typeof DataEncoding)[keyof typeof DataEncoding];

// ── Intent sub-records (1:1 byte layout with the Rust structs) ────────
//
// These appear inside a parsed `IntentAccount.params / accounts /
// instructions / data_segments / seeds` array. Keep fields in the same
// order as the Rust structs for the sake of readers; no byte layout is
// implied here (the account parser emits these records fully decoded).

export interface ParamEntry {
  paramType: ParamType;
  nameOffset: number;
  nameLen: number;
  constraintType: ConstraintType;
  constraintValue: bigint;
}

export interface AccountEntry {
  isSigner: boolean;
  isWritable: boolean;
  sourceType: AccountSourceType;
  poolOffset: number;
  poolLen: number;
}

export interface InstructionEntry {
  programAccountIndex: number;
  accountIndexesOffset: number;
  accountIndexesLen: number;
  segmentsStart: number;
  segmentsCount: number;
}

export interface DataSegmentEntry {
  segmentType: SegmentType;
  poolOffset: number;
  poolLen: number;
}

export interface SeedEntry {
  seedType: SeedType;
  poolOffset: number;
  poolLen: number;
}

/// Variable-width param size in `params_data`. Matches
/// programs/clear-wallet/src/utils/definition.rs::param_byte_size.
export function paramByteSize(
  paramType: ParamType,
  paramsData: Uint8Array,
  offset: number
): number {
  switch (paramType) {
    case ParamType.Address:
    case ParamType.Bytes32:
      return 32;
    case ParamType.U64:
    case ParamType.I64:
      return 8;
    case ParamType.Bytes20:
      return 20;
    case ParamType.String: {
      if (offset >= paramsData.length) {
        throw new Error("paramByteSize: string length byte OOB");
      }
      return 1 + paramsData[offset];
    }
    case ParamType.Bytes: {
      if (offset + 1 >= paramsData.length) {
        throw new Error("paramByteSize: bytes length OOB");
      }
      const len = paramsData[offset] | (paramsData[offset + 1] << 8);
      return 2 + len;
    }
    case ParamType.Bool:
    case ParamType.U8:
      return 1;
    case ParamType.U16:
      return 2;
    case ParamType.U32:
      return 4;
    case ParamType.U128:
      return 16;
    default: {
      const exhaustive: never = paramType;
      throw new Error(`paramByteSize: unknown ParamType ${exhaustive}`);
    }
  }
}

/// Byte offset of the `idx`-th param inside `paramsData`. Walks the
/// preceding params using `paramByteSize`. Used by the renderer + the
/// message builder to pick out individual fields from the opaque blob.
export function paramOffsetAt(
  params: readonly { paramType: ParamType }[],
  paramsData: Uint8Array,
  idx: number
): number {
  let off = 0;
  for (let i = 0; i < idx; i++) {
    const p = params[i];
    if (!p) throw new Error(`paramOffsetAt: param index ${i} out of bounds`);
    off += paramByteSize(p.paramType, paramsData, off);
  }
  return off;
}
